docs/source/modernizing.rst

```diff
@@ -0,0 +1,319 @@
+Modernizing Obsolete Typing Features
+************************************
+
+Introduction
+============
+
+This guide helps to modernize your code by replacing older typing features
+with their modern equivalents. Not all features described here are obsolete,
+but they superseded by more modern alternatives, that are recommended to use.
+
+These newer features are not available in all Python versions, although
+some features are available as backports from the
+`typing-extensions <https://pypi.org/project/typing-extensions/>`_
+package, or require quoting or using :python:`from __future__ import annotations`.
+Each section states the minimum Python version required to use the
+feature, whether it is available in typing-extensions, and whether it is
+available using quoting.
+
+.. note::
+
+    The latest version of typing-extensions is available for all Python
+    versions that have not reached their end of life, but not necessarily for
+    older versions.
+
+.. note::
+
+    :python:`from __future__ import annotations` is available since Python 3.7.
+    This will only work inside type annotations, while quoting is still
+    required outside. For example, this example runs on Python 3.7 and up,
+    although the pipe operator was only introduced in Python 3.10::
+
+        from __future__ import annotations
+        from typing_extensions import TypeAlias
+
+        def f(x: int | None) -> int | str: ...  # the future import is sufficient
+        Alias: TypeAlias = "int | str"  # this requires quoting
+
+.. _modernizing-type-comments:
+
+Type Comments
+=============
+
+*Alternative available since:* Python 3.0, 3.6
+
+Type comments were originally introduced to support type annotations in
+Python 2 and variable annotations before Python 3.6. While most type checkers
+still support them, they are considered obsolete, and type checkers are
+not required to support them.
+
+For example, replace::
+
+    x = 3  # type: int
+    def f(x, y):  # type: (int, int) -> int
+        return x + y
+
+with::
+
+    x: int = 3
+    def f(x: int, y: int) -> int:
+        return x + y
+
+When using forward references or types only available during type checking,
+it's necessary to either use :python:`from __future__ import annotations`
+(available since Python 3.7) or to quote the type::
+
+    def f(x: "Parrot") -> int: ...
+
+    class Parrot: ...
+
+.. _modernizing-typing-text:
+
+``typing.Text``
+===============
+
+*Alternative available since:* Python 3.0
+
+:class:`typing.Text` was a type alias intended for Python 2 compatibility.
+It is equivalent to :class:`str` and should be replaced with it.
+For example, replace::
+
+    from typing import Text
+
+    def f(x: Text) -> Text: ...
+
+with::
+
+    def f(x: str) -> str: ...
+
+.. _modernizing-typed-dict:
+
+``typing.TypedDict`` Legacy Forms
+=================================
+
+*Alternative available since:* Python 3.6
+
+:class:`TypedDict <typing.TypedDict>` supports two legacy forms for
```

Comment:
We can add that the first legacy form is still useful if you have keys that are not valid identifiers.